### ISSUE TITLE
Is there any way to return a boolean from assertions?

### ISSUE THREAD

**Issue by raschmitt** [fluentassertions/fluentassertions#1498]
Is there any way to return a boolean from assertions?

Something in these lines:

```csharp
bool result = 2.Should().Be(1)
```

Where bool would be returned as `false` in this case.

**Comment by jnyrup** [comment]
The closest you can get is to create a new `AssertionScope` and inspects _its_ state after invoking the assertion.

```cs
var scope = new FluentAssertions.Execution.AssertionScope();
2.Should().Be(1);
var result = scope.Succeeded;
```

Out of curiosity, what are you trying to achieve?
I'm also asking because you there might be a better solution, if you can provide the broader picture.

**Comment by raschmitt** [comment]
I'm trying to use `fluentAssertions` to assert parameters inside an `Arg.Is<T>` of `Nsubstitute`, where I need to 

https://nsubstitute.github.io/help/argument-matchers/

Something in the lines as discussed here: https://github.com/fluentassertions/fluentassertions/issues/158

**Comment by jnyrup** [comment]
You could use the method proposed in https://github.com/fluentassertions/fluentassertions/issues/1402#issuecomment-704031328
```cs
public static AssertionHelper
{
    public static bool ToPredicate(Action act)
    {
        using (var scope = new AssertionScope())
        {
            act();
            return scope.Discard().Any();
        }
    }
}
```

and then simplify the calling code a bit with `using static`
```cs
using static AssertionHelper;

result = Assert(() => 2.Should().Be(1))
```

**Comment by raschmitt** [comment]
Yes, I was doing something similar, it works. Yet I think it is very "ugly" to use with a fluent library.

I wish we could use an extension method, but then on what? As assertions return void results.

**Comment by dennisdoomen** [comment]
It's an assertion library for use in unit tests, not a validation library for use in production code. What you're looking for sounds more like [FluentValidation](https://fluentvalidation.net/).

Either way, this is the only workaround we have right now.

**Comment by Fed03** [comment]
With NSubsitute I came with the following solution:
```c#
internal class CapturedArg<T>
{
    private readonly T _capturedArg;

    public CapturedArg(T capturedArg)
    {
        _capturedArg = capturedArg;
    }

    public void ShouldSatisfy(Action<T> conditions)
    {
        conditions(_capturedArg);
    }
}

[SuppressMessage("ReSharper", "NS1004")]
private static ref T CaptureArg<T>(out Task<CapturedArg<T>> arg, int timeoutMs = 10000)
{
    var source = new TaskCompletionSource<CapturedArg<T>>();
    arg = source.Task;
    var ct = new CancellationTokenSource(timeoutMs);
    ct.Token.Register(
        () => source.TrySetException(
            new TimeoutException(
                $"Cannot capture the arg because the method has not been called within the allowed time span. Check that the method is being called or increase the timeout."
            )
        ),
        false
    );

    return ref Arg.Do<T>(innerArg => source.TrySetResult(new CapturedArg<T>(innerArg)));
}
```
that can be used like this
```c#
mock.method(CaptureArg<type>(out var capturedArg)).returns(something);

sut.doSomething();

(await capturedArg).ShouldSatisfy(arg => {
    arg.Property1.Should().Be("foo");
    arg.Property2.Should().Be("bar");
});
```

**Comment by megasuperlexa** [comment]
It is a pity this did not get attention, because when I am inside an AssertionScope, the assertion might fail without exception, and the code that follows will be pointlessly executed.  
For example I assert something to be not null and then proceed to some logic and eventually access properties of that 'something' for further assertions.
That will produce ugly nullref output despite check being in place.

**Comment by dennisdoomen** [comment]
> That will produce ugly nullref output despite check being in place.

Are you referring to the IDE warning about possible null references? Because should have been addressed by #2380
